File: 102-binary-tree-level-order-traversal/dennis/102.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
import Queue

class Solution(object):
    def levelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """

        q = Queue.Queue()
        if None != root:
            q.put(root)

        final_ans = []

        while not q.empty():
            num_of_node_this_round = q.qsize()
            item_this_round = []
            for i in xrange(num_of_node_this_round):
                node = q.get()
                item_this_round.append(node.val)
                if None != node.left:
                    q.put(node.left)
                if None != node.right:
                    q.put(node.right)
            final_ans.append(item_this_round)

        return final_ans

File: 102-binary-tree-level-order-traversal/leochang/102.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	vector<vector<int>> levelOrder(TreeNode* root) {
		vector<vector<int> >  vResult;
		vector<int> vRow;
		queue<TreeNode*> qInProcess;
		qInProcess.push(root);
		int size = 1;

		if(NULL == root) return vResult;

		while(!qInProcess.empty()) {
			if (qInProcess.front()->left) {
				qInProcess.push(qInProcess.front()->left);
			}
			if (qInProcess.front()->right) {
				qInProcess.push(qInProcess.front()->right);
			}

			vRow.emplace_back(qInProcess.front()->val);
			qInProcess.pop();
			size--;

			if (0 == size) {
				vResult.push_back(vRow);
				vRow.clear();
				size = qInProcess.size();
			}
		}

		return vResult;
	}
};
